### PR TITLE
[W-8791351] - Create Disbursements component permissions

### PR DESCRIPTION
**Pull request by jdavilasf**
# Critical Changes

# Changes
- `DisbursementsCreate` aura component was updated to inform when the user do not have access to  create a `Disbursement` record.
- `disbursementsCreateController` was updated to validate user access before inserting `Disbursement` record.  
- Modified `force-app/main/default/classes` directory structure to better contain related Apex classes.  

# Issues Closed
- [Create Disbursements component does not honour user object permissions for the Disbursement object
](https://github.com/SalesforceFoundation/OutboundFunds/issues/78)

# New Metadata
- `FundingRequestSelector` Apex class that enforce read object-level and field-level access on queries used on `disbursementsCreateController`
- `DatabaseService` Apex class checks for object-level access before performing database operation.
- `UnitTest` Apex class unit test framework.
- `TestUser` Apex class to help test permissions.

# Deleted Metadata

# Definition of Done
Refer to [Definition of Done](https://salesforce.quip.com/9P7hAOPHJJyU) to see any additional details for the items below:
- [x] ~~Any net new LWC work has JEST test coverage 50% or above~~
- [x] ~~Default Sa11y tests pass for all LWC components~~
- [x] 🔒 Secure both Front-end (LWC) & back-end (Apex) as necessary
- [x] 🔑 ~~Grant users access in Permission Sets (Object, Field, Apex Class) as necessary~~
- [x] Link the pull request and work item by PR comment and Chatter post respectively, e.g. GUS: [W-0000000: Work Name]()
- [x] Make sure that ACs are updated (if any gaps)
- [ ] **All acceptance criteria have been met**
    - [x] Developer
    - [x] Code Reviewer
- [x] Pull Request contains draft release notes
- [x] Labels, help text, and customer facing messages are reviewed by Docs
- [x] QE story level testing completed

### CODE REVIEW

**Comment by jdavilasf** [comment]
[W-8791351](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07B00000094gliIAA/view)

File: force-app/main/default/classes/selectors/tests/FundingRequestSelector_TEST.cls [hunk 1]
```diff
@@ -0,0 +1,64 @@
+@isTest
```
Comment (scottwarren-sfdo) [inline force-app/main/default/classes/selectors/tests/FundingRequestSelector_TEST.cls:1]: Can we add tests to verify that object and field permissions are respected in the selector? (Scott P or I can help you with this in the morning)

**Review by scottwarren-sfdo — CHANGES_REQUESTED** [review]
Great work Julio -- I've added some comments. We can discuss / look at them together tomorrow. 

When you have a moment, please also take a look at the PR's DoD items. I only see one box checked atm.

File: force-app/main/default/aura/disbursementsCreate/disbursementsCreateHelper.js [hunk 2]
```diff
@@ -132,22 +132,35 @@
...
-            // Clear these out after saved
-            cmp.set("v.model.disbursements", null);
 
-            // Refresh Record Page
-            $A.get("e.force:refreshView").fire();
+        let saveButton = event.getSource();
+        saveButton.set("v.disabled", true);
 
-            $A.get("e.force:closeQuickAction").fire();
+        this.callServer(cmp, "c.saveDisbursements", params, function (r) {
```
[the remark this answers is not in the document]
Comment (jdavilasf) [reply force-app/main/default/aura/disbursementsCreate/disbursementsCreateHelper.js:146]: Agree. I'm switching to `result`. to make reference to the return value from the callback.

File: force-app/main/default/aura/disbursementsCreate/disbursementsCreateHelper.js [hunk 3]
```diff
@@ -132,22 +132,35 @@
...
+            let userHasAccess = r;
```
[the remark this answers is not in the document]
Comment (jdavilasf) [reply force-app/main/default/aura/disbursementsCreate/disbursementsCreateHelper.js:147]: I tried to make it more descriptive. I'm switching to `result` directly instead.

File: force-app/main/default/aura/disbursementsCreate/disbursementsCreateHelper.js [hunk 4]
```diff
@@ -233,6 +246,8 @@
         this.validateAmountRemaining(cmp);
 
         this.validateTotal(cmp);
+
+        //
```
Comment (scottwarren-sfdo) [inline force-app/main/default/aura/disbursementsCreate/disbursementsCreateHelper.js:250]: Can you please remove these extra lines?

File: force-app/main/default/classes/selectors/tests/FundingRequestSelector_TEST.cls [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+                Schema.Funding_Request__c.SObjectType.fields.Total_Remaining__c.getDescribe()
+                    .isAccessible()
+            );
+
+            System.assertEquals(
+                false,
+                Schema.Funding_Request__c.SObjectType.fields.Total_Remaining__c.getDescribe()
+                    .isAccessible()
+            );
+            // Question regarding lookups
```
Comment (scottwarren-sfdo) [inline force-app/main/default/classes/selectors/tests/FundingRequestSelector_TEST.cls:120]: I believe we handled this question earlier? Let me know if there are more!

File: force-app/main/default/classes/selectors/tests/FundingRequestSelector_TEST.cls [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+
+            // Run the test
+            String fundingRequestId = [SELECT Id FROM Funding_Request__c LIMIT 1].Id;
```
Comment (scottwarren-sfdo) [inline force-app/main/default/classes/selectors/tests/FundingRequestSelector_TEST.cls:123]: Shouldn't we be retrieving the ID outside of the run-as block? I wouldn't expect this query to return a record as the min-access persona.

Comment (jdavilasf) [reply force-app/main/default/classes/selectors/tests/FundingRequestSelector_TEST.cls:123]: That makes total sense let me move it up

File: force-app/main/default/classes/selectors/tests/FundingRequestSelector_TEST.cls [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+
+        final FundingRequestSelector fundingRequestSelector = new FundingRequestSelector();
+
+        List<Funding_Request__c> fundingRequests = fundingRequestSelector.getFundingRequestById(
+            fundingRequestId
+        );
+
+        Test.stopTest();
+
+        System.assertNotEquals(
```
Comment (scottwarren-sfdo) [inline force-app/main/default/classes/selectors/tests/FundingRequestSelector_TEST.cls:189]: Do you mean to compare a Null with a Boolean here? Could we `assertNotEquals` `True` with `isEmpty()` ?

Comment (jdavilasf) [reply force-app/main/default/classes/selectors/tests/FundingRequestSelector_TEST.cls:189]: Good catch... I'm updating it now

**Review by scottwarren-sfdo — APPROVED** [review]
Thanks for all of the updates! This is ready for testing now.

File: force-app/main/default/labels/CustomLabels.labels-meta.xml [hunk 8]
```diff
@@ -1,180 +1,203 @@
...
-        <fullName>disbursementsCreate_CalculateDisbursements</fullName>
-        <categories>Aura</categories>
+        <fullName>DatabaseService_NoInsertFlsException</fullName>
+        <categories>DatabaseService,Exception</categories>
         <language>en_US</language>
         <protected>true</protected>
-        <shortDescription>disbursementsCreate: Calculate Disbursements</shortDescription>
-        <value>Calculate Disbursements</value>
+        <shortDescription>The user does not have create access to fields required for this process.</shortDescription>
+        <value>You don&apos;t have field-level access to create this record. Contact your system administrator for help.</value>
```
Comment (nicolersmith) [inline force-app/main/default/labels/CustomLabels.labels-meta.xml:25]:
```suggestion
        <value>You don&apos;t have field-level access to create this record. Ask your Salesforce admin for help.</value>
```

File: force-app/main/default/labels/CustomLabels.labels-meta.xml [hunk 9]
```diff
@@ -1,180 +1,203 @@
...
+    </labels>  
+    <labels>
+        <fullName>DatabaseService_NoUpdateFlsException</fullName>
+        <categories>DatabaseService,Exception</categories>
+        <language>en_US</language>
+        <protected>true</protected>
+        <shortDescription>The user does not have edit or update access to required fields.</shortDescription>
+        <value>You don&apos;t have field-level access to edit this record. Contact your system administrator for help.</value>
```
Comment (nicolersmith) [inline force-app/main/default/labels/CustomLabels.labels-meta.xml:33]:
```suggestion
        <value>You don&apos;t have field-level access to edit this record. Ask your Salesforce admin for help.</value>
```

**Review by nicolersmith — APPROVED** [review]

**Pull request merged**